nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesInformationBuilder.java

```diff
@@ -0,0 +1,140 @@
+import static org.sonatype.nexus.repository.r.internal.RAttributes.P_VERSION;
+
+/**
+ * Builds the contents of a PACKAGES.gz file based on the provided assets, taking into account the greatest version of a
+ * particular package that is available in a (hosted) repository.
+ *
+ * Note that this maintains all pertinent information for the "latest" version of each package in memory, though the
+ * actual amount of information for each package is rather small.
+ *
+ * TODO: Add support for other metadata types (PACKAGES, PACKAGES.rds etc.)
+ */
+public class RPackagesInformationBuilder
+{
+  /**
+   * The greatest version of each package currently encountered during this run.
+   */
+  private final Map<String, RPackageVersion> packageVersions = new HashMap<>();
+
+  /**
+   * The package information as of the last time we processed the greatest version of this package. A {@code TreeMap} is
+   * used to maintain ordering by package name.
+   */
+  private final Map<String, Map<String, String>> packageInformation = new TreeMap<>();
+
+  /**
+   * Processes an asset, updating the greatest version and details for the package if appropriate.
+   *
+   * @param asset The asset to process.
+   */
+  public void append(final Asset asset) {
+    // is this a newer version of this asset's package than the one we currently have (if we have one)?
+    String packageName = asset.formatAttributes().get(P_PACKAGE, String.class);
+    RPackageVersion oldVersion = packageVersions.get(packageName);
+    RPackageVersion newVersion = new RPackageVersion(asset.formatAttributes().get(P_VERSION, String.class));
+    if (oldVersion == null || newVersion.compareTo(oldVersion) > 0) {
+
+      // if so, use the most recent information instead and update the greatest version encountered
+      Map<String, String> newInformation = new HashMap<>();
+      newInformation.put(P_PACKAGE, asset.formatAttributes().get(P_PACKAGE, String.class));
+      newInformation.put(P_VERSION, asset.formatAttributes().get(P_VERSION, String.class));
+      newInformation.put(P_DEPENDS, asset.formatAttributes().get(P_DEPENDS, String.class));
+      newInformation.put(P_IMPORTS, asset.formatAttributes().get(P_IMPORTS, String.class));
+      newInformation.put(P_SUGGESTS, asset.formatAttributes().get(P_SUGGESTS, String.class));
+      newInformation.put(P_LICENSE, asset.formatAttributes().get(P_LICENSE, String.class));
+      newInformation.put(P_NEEDS_COMPILATION, asset.formatAttributes().get(P_NEEDS_COMPILATION, String.class));
+
+      packageVersions.put(packageName, newVersion);
+      packageInformation.put(packageName, newInformation);
+    }
+  }
+
+  /**
+   * Using collected package details builds PACKAGES.gz file and returns it as byte array.
+   * <p>
+   * Call this method ONLY after all information about packages is appended to packageInformation map.
+   *
+   * @return PACKAGES.gz as byte array.
+   */
+  public byte[] buildPackagesGz() throws IOException {
```

Comment:
Too big complexity, please split this method to decrease deep of structure

Reply:
done in fe40919080a00ea9162d0af9be02c414af338e27